File: architecture/data/src/main/java/com/pacific/data/SystemIo.java
package com.pacific.data;

import android.content.Context;
import android.content.Intent;
import android.content.res.AssetManager;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Environment;
import android.provider.MediaStore;
import android.support.annotation.WorkerThread;
import android.text.TextUtils;
import android.util.Log;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;
import okhttp3.internal.Util;
import okhttp3.internal.io.FileSystem;
import okio.BufferedSink;
import okio.BufferedSource;
import okio.Okio;

public class SystemIo {

  private final static String TAG = "SystemIo";

  private SystemIo() {
    throw new UnsupportedOperationException();
  }

  @WorkerThread
  public static void readAsset(Context context, String name, String path, boolean overwrite) {
    DataUtil.verifyWorkThread();
    path = suffixDir(path);
    mkdirs(path);
    AssetManager assetManager = context.getAssets();
    try {
      final File file = new File(path + name);
      if (FileSystem.SYSTEM.exists(file)) {
        if (overwrite) {
          FileSystem.SYSTEM.delete(file);
        } else {
          return;
        }
      }
      file.createNewFile();
      BufferedSource source = Okio.buffer(Okio.source(assetManager.open(name)));
      BufferedSink sink = Okio.buffer(FileSystem.SYSTEM.sink(file));
      byte[] bytes = new byte[1024];
      int nRead;
      while ((nRead = source.read(bytes)) != -1) {
        sink.write(bytes, 0, nRead);
      }
      Util.closeQuietly(source);
      Util.closeQuietly(sink);
    } catch (IOException e) {
      Log.e(TAG, e.toString());
    }
  }

  @WorkerThread
  public static void toGallery(Context context, Bitmap bmp, File directory, String img) {
    DataUtil.verifyWorkThread();
    final String ext = ".jpg";
    if (!FileSystem.SYSTEM.exists(directory) || !directory.isDirectory()) {
      directory.mkdir();
    }
    if (TextUtils.isEmpty(img)) {
      img = System.currentTimeMillis() + ext;
    }
    if (!img.endsWith(ext)) {
      img = img + ext;
    }
    File file = new File(directory, img);
    try {
      if (FileSystem.SYSTEM.exists(file)) {
        FileSystem.SYSTEM.delete(file);
      }
      file.createNewFile();
      FileOutputStream fos = new FileOutputStream(file);
      bmp.compress(Bitmap.CompressFormat.JPEG, 100, fos);
      fos.flush();
      Util.closeQuietly(fos);
    } catch (FileNotFoundException e) {
      Log.e(TAG, e.toString());
    } catch (IOException e) {
      Log.e(TAG, e.toString());
    }
    try {
      MediaStore
          .Images
          .Media
          .insertImage(context.getContentResolver(), file.getAbsolutePath(), img, null);
      context.sendBroadcast(new Intent(Intent.ACTION_MEDIA_SCANNER_SCAN_FILE, Uri.fromFile(file)));
    } catch (FileNotFoundException e) {
      Log.e(TAG, e.toString());
    }
  }

  @WorkerThread
  public static void unzip(File zipFile, String directory) {
    DataUtil.verifyWorkThread();
    try {
      ZipInputStream zipInputStream = new ZipInputStream(new FileInputStream(zipFile));
      BufferedSource source = Okio.buffer(Okio.source(zipInputStream));
      directory = suffixDir(directory);
      ZipEntry zipEntry;
      BufferedSink sink;
      while ((zipEntry = zipInputStream.getNextEntry()) != null) {
        if (zipEntry.isDirectory()) {
          mkdirs(directory + File.separator + zipEntry.getName());
        } else {
          sink = Okio.buffer(Okio.sink(new File(directory, zipEntry.getName())));
          byte[] bytes = new byte[1024];
          int nRead;
          while ((nRead = source.read(bytes)) != -1) {
            sink.write(bytes, 0, nRead);
          }
          Util.closeQuietly(sink);
          zipInputStream.closeEntry();
        }
      }
      Util.closeQuietly(source);
      Util.closeQuietly(zipInputStream);
    } catch (IOException e) {
      Log.e(TAG, e.toString());
    }
  }

  public static String getSDCard() {
    return suffixDir(Environment.getExternalStorageDirectory().getAbsolutePath());
  }

  @WorkerThread
  public static File mkdirs(String dir) {
    DataUtil.verifyWorkThread();
    File file = new File(dir);
    if (!file.exists() || !file.isDirectory()) {
      file.mkdirs();
    }
    return file;
  }

  public static String suffixDir(String dir) {
    if (dir.endsWith(File.separator)) {
      return dir;
    }
    return dir + File.separator;
  }
}

File: pacific-architecture/guava-android/src/main/java/com/pacific/guava/android/io/SystemIO.kt
package com.pacific.guava.android.io

import android.content.Context
import androidx.annotation.WorkerThread
import com.pacific.guava.android.ensureWorkThread
import com.pacific.guava.jvm.io.ensureFileSeparator
import com.pacific.guava.jvm.io.mkdirs
import okhttp3.internal.closeQuietly
import okhttp3.internal.io.FileSystem
import okio.buffer
import okio.source
import java.io.File
import java.io.IOException

/**
 * 复制文件
 *
 * @param targetDir 目标目录
 * @param file assert资源文件名字
 * @param overwrite 是否覆盖（如果需要）
 */
@WorkerThread
fun copy(targetDir: String, file: String, overwrite: Boolean): File? {
    return copy(targetDir, File(file), overwrite)
}

/**
 * 复制文件
 *
 * @param targetDir 目标目录
 * @param file assert资源文件名字
 * @param overwrite 是否覆盖（如果需要）
 */
@WorkerThread
fun copy(targetDir: String, file: File, overwrite: Boolean): File? {
    ensureWorkThread()
    try {
        val file = File(ensureFileSeparator(mkdirs(targetDir).absolutePath) + file.name)
        if (FileSystem.SYSTEM.exists(file)) {
            if (overwrite) {
                FileSystem.SYSTEM.delete(file)
            } else {
                return file
            }
        }
        file.createNewFile()
        val source = file.source().buffer()
        val sink = FileSystem.SYSTEM.sink(file).buffer()
        val bytes = ByteArray(1024)
        var nRead: Int = source.read(bytes)
        while (nRead != -1) {
            sink.write(bytes, 0, nRead)
            nRead = source.read(bytes)
        }
        source.closeQuietly()
        sink.closeQuietly()
        return file
    } catch (e: IOException) {
        e.printStackTrace()
        return null
    }
}

/**
 * 复制assert文件
 *
 * @param context 目标目录
 * @param name assert资源文件名字
 * @param targetDir 是否覆盖（如果需要）
 * @param overwrite 是否覆盖（如果需要）
 */
@WorkerThread
fun copyFromAsset(context: Context, name: String, targetDir: String, overwrite: Boolean) {
    ensureWorkThread()
    try {
        val file = File(ensureFileSeparator(mkdirs(targetDir).absolutePath) + name)
        if (FileSystem.SYSTEM.exists(file)) {
            if (overwrite) {
                FileSystem.SYSTEM.delete(file)
            } else {
                return
            }
        }
        file.createNewFile()
        val source = context.assets.open(name).source().buffer()
        val sink = FileSystem.SYSTEM.sink(file).buffer()
        val bytes = ByteArray(1024)
        var nRead: Int = source.read(bytes)
        while (nRead != -1) {
            sink.write(bytes, 0, nRead)
            nRead = source.read(bytes)
        }
        source.closeQuietly()
        sink.closeQuietly()
    } catch (e: IOException) {
        e.printStackTrace()
    }
}

/**
 * 读取assert文件内容
 *
 * @param context 上下文
 * @param name assert资源文件名字
 */
@WorkerThread
fun readFromAsset(context: Context, name: String): String {
    ensureWorkThread()
    val assetManager = context.assets
    var content = ""
    try {
        val source = assetManager.open(name).source().buffer()
        content = source.readUtf8()
        source.closeQuietly()
    } catch (e: IOException) {
        e.printStackTrace()
    }
    return content
}